aws/resource_aws_db_parameter_group.go

```diff
@@ -290,8 +290,26 @@ func resourceAwsDbParameterGroupUpdate(d *schema.ResourceData, meta interface{})
 			}
 		}
 
+		// Build a map with names from the new parameter set
+		newNames := make(map[string]bool, ns.Len())
```

Comment:
Shouldn't we move the removed / reset logic into the above modify loop so there is one place where a parameter change is managed? That way we can:

```
for , parameter := range parameters {
  if parameter in ns and not in os {
   add
  } else if parameter in ns and in os {
   modify
 } else if paremeter not in ns and in os {
  delete / reset to default
 }
```

The way we have the code now, and after this change, acts on modified values. The bug is that `parameters, err := expandParameters(ns.Difference(os).List())` generates the same set of values as `resetParameters, err := expandParameters(os.Difference(ns).List())`.

Reply:
Sorry, I was wrong, this is a set difference. Order matters - ns.Difference(os) is the values in ns that are not in os.

Reply:
Yeah, the parameters that have been removed shouldn't be caught by the `parameters, err := expandParameters(ns.Difference(os).List())` call which precedes the block which processes modified parameters since none of the removed parameters are in `ns`. As @lachlancooper explained really well in https://github.com/terraform-providers/terraform-provider-aws/issues/11846#issuecomment-586777896, the `resetParameters, err := expandParameters(os.Difference(ns).List())` call would catch any parameters whose name or other attributes (including value) appear in `os` but not in `ns`. This leads to parameters which have the same name in `os` and `ns` but different values getting unintentionally reset, which is what this PR intends to fix.